### DISCUSSION TITLE
How to type the prisma client

### DISCUSSION THREAD

**Discussion by jakeleventhal** [prisma/prisma#18502] [Q&A] [answered]
What is the proper type to use in this case for `transaction` for instance in the function stub for `createAppData`
<img width="568" alt="Screenshot 2023-03-17 at 10 18 48 PM" src="https://user-images.githubusercontent.com/15133828/227732811-1d885293-eb1a-4099-ad09-c4b5794c0e39.png">

**Comment by nurul3101** [comment, marked as answer]
Hello @jakeleventhal 👋 

Can you please try using the type `Prisma.TransactionClient` like this?

```ts
import { PrismaClient, Prisma } from '@prisma/client';

const prisma = new PrismaClient({
  log: ['query'],
});

async function main() {
  await prisma.$transaction(
    async (transaction: Prisma.TransactionClient) => {}
  );
}

main()
  .catch((e) => {
    throw e;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
```

**Reply by jakeleventhal** [reply]
This seems to work; however, it doesn't account for any $extends behavior such as what i have
```typescript
new PrismaClient().$extends({
  model: {
    $allModels: {
      async exists<T>(this: T, where: Prisma.Args<T, 'findFirst'>['where']): Promise<boolean> {
        const context = Prisma.getExtensionContext(this);
        const result = await (context as any).count({ where });
        return result > 0;
      }
    }
  }
});
```

**Reply by jakeleventhal** [reply]
it says `exists` doesn't exist on `Prisma.TransactionClient`

**Reply by nurul3101** [reply]
Do you mind sharing this feedback in this [GitHub Issue](https://github.com/prisma/prisma/issues/16500) so that our Engineering team working on extensions can have a look and share their insights?

**Reply by jakeleventhal** [reply]
shared feedback there

**Reply by nurul3101** [reply]
Thank you so much! 🙌

**Reply by jakeleventhal** [reply]
Following up:

I was able to create a type like this
```ts
const client = () =>
  new PrismaClient().$extends({
    model: {
      $allModels: {
        async exists<T>(this: T, where: Prisma.Args<T, 'findFirst'>['where']): Promise<boolean> {
          const context = Prisma.getExtensionContext(this);
          const result = await (context as any).count({ where });
          return result > 0;
        }
      }
    }
  });

export type ExtendedPrismaClient = ReturnType<typeof client>;
```

**Reply by nurul3101** [reply]
Thank you so much for sharing it here! 🚀
